xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-solr/xwiki-platform-search-solr-api/src/main/java/org/xwiki/search/solr/internal/metadata/DefaultLinkStore.java

```diff
@@ -0,0 +1,244 @@
+    {
+        try {
+            return this.solr.getClient(SolrClientInstance.CORE_NAME);
+        } catch (SolrException e) {
+            throw new LinkException("Failed to acces Solr search core", e);
+        }
+    }
+
+    @Override
+    @SuppressWarnings("resource")
+    public Set<EntityReference> resolveLinkedEntities(EntityReference target) throws LinkException
+    {
+        // Get the id of the reference
+        String solrID;
+        try {
+            solrID = this.searchUtils.getId(target);
+        } catch (Exception e) {
+            throw new LinkException("Failed to serialize the Solr id for the reference [" + target + "]", e);
+        }
+
+        // Load the Solr document corresponding to the reference
+        SolrDocument solrDocument;
+        try {
+            solrDocument = getClient().getById(solrID);
+        } catch (Exception e) {
+            throw new LinkException(
+                "Failed to load the Solr document for the reference [" + target + "] (id [" + solrID + "])", e);
+        }
+
+        if (solrDocument == null) {
+            return Collections.emptySet();
+        }
+
+        // Get the links from the solr document
+        Collection<Object> links = solrDocument.getFieldValues(FieldUtils.LINKS);
+        Set<EntityReference> entities = new HashSet<>(links.size());
+        for (Object link : links) {
+            if (link instanceof String) {
+                EntityReference entityLink = this.linkSerializer.unserialize((String) link);
+
+                if (entityLink != null) {
+                    // Make sure to resolve the reference as a DOCUMENT based references and not a PAGE one
+                    EntityReference documentLink = toDocumentBasedReference(entityLink);
+
+                    entities.add(documentLink);
+                }
+            }
+        }
+
+        return entities;
+    }
+
+    @Override
+    public Set<EntityReference> resolveBackLinkedEntities(EntityReference reference) throws LinkException
+    {
+        // Get the PAGE based reference
+        EntityReference pageBasedReference = toPageBasedReference(reference);
+        // Get the DOCUMENT based reference
+        EntityReference documentBasedReference = toDocumentBasedReference(reference);
+
+        // Search for the Solr entities which contains extended links with the resolved DOCUMENT and PAGE references
+
+        // Get only the entities with links to the passed reference (either as DOCUMENT or PAGE version)
+        StringBuilder filter = new StringBuilder();
+        if (pageBasedReference != null) {
+            filter.append(FieldUtils.LINKS_EXTENDED);
+            filter.append(':');
+            filter.append(this.utils.toCompleteFilterQueryString(this.linkSerializer.serialize(pageBasedReference)));
+        }
+        if (filter.length() > 0) {
+            filter.append(" OR ");
+        }
+        filter.append(FieldUtils.LINKS_EXTENDED);
+        filter.append(':');
```

Comment:
Two occurrences of `:` in same file? I'm surprised checkstyle didn't complain.

Reply:
Checkstyle does not care about characters, it only checks strings.

Reply:
right